File: src/chanhyun/week3/표 편집.py
def solution(n, k, cmd):
    answer = ''
    graph = {i: [i - 1, i + 1] for i in range(0, n)}
    graph[0] = [n-1, 1]
    graph[n-1] = [n-2, 0]
    stack = []
    ptr = k
    for c in cmd:
        if c == 'C':
            x, y = graph[ptr]
            graph[x][1] = y
            graph[y][0] = x
            
            stack.append([ptr, x, y])
            del graph[ptr]
            
            if y < ptr:
                ptr = x
            else:
                ptr = y
        elif c == 'Z':
            p, x, y = stack.pop()
            graph[p] = [x,y]
            graph[x][1] = p
            graph[y][0] = p
        else:
            o, cnt = c.split()
            if o == 'U':
                for _ in range(int(cnt)):
                    ptr = graph[ptr][0]
            else:
                for _ in range(int(cnt)):
                    ptr = graph[ptr][1]
    
    for i in range(n):
        if i in graph:
            answer += 'O'
        else:
            answer += 'X'
    return answer

File: src/taeeun/week3/표 편집.js
function solution(n, k, cmd) {
    let answer=new Array(n).fill('O');
    
    const Node = function(idx, prev) {
        this.idx=idx;
        this.prev=prev;
        this.next=null;
    }
    
    let prevNode= new Node(0);
    let current=prevNode;
    
    //연결리스트 생성
    for (let i=1; i<n; i++) {
        const nextNode = new Node(i, prevNode);
        prevNode.next=nextNode;
        prevNode=nextNode;
        if (i===k) current=nextNode;
    }
    
    const deletedNodes=[];
    
    const deleteNode=()=>{
        const prev=current.prev;
        const next=current.next;
        
        deletedNodes.push(current);
        
        current = next ? next : prev;
        if (prev) prev.next=next;
        if (next) next.prev=prev;
    }
    
    const recoverNode=()=>{
        const recoveredNode = deletedNodes.pop();
        const prev = recoveredNode.prev;
        const next = recoveredNode.next;
        
        if (prev) prev.next=recoveredNode;
        if (next) next.prev=recoveredNode;
    }
    
    for (const c of cmd) {
        const [command, num] = c.split(' ');
        let i=0;
        switch(command) {
            case 'U':
                while (i<num && current.prev) {
                    i++;
                    current=current.prev;
                }
                break;
            case 'D':
                while (i<num && current.next) {
                    i++;
                    current=current.next;
                }
                break;
            case 'Z':
                recoverNode();
                break;
            case 'C': 
                deleteNode(); 
                break;
        }
    }
    
    deletedNodes.map((node)=>answer[node.idx]='X');

    return answer.join('');
}
